x-pack/plugins/security/public/management/roles/indices_api_client.ts

```diff
@@ -8,12 +8,25 @@
 import type { HttpStart } from 'src/core/public';
 
 export class IndicesAPIClient {
+  private fieldCache = new Map<string, string[]>();
+
   constructor(private readonly http: HttpStart) {}
 
-  async getFields(query: string) {
-    return (
-      (await this.http.get<string[]>(`/internal/security/fields/${encodeURIComponent(query)}`)) ||
-      []
-    );
+  async getFields(pattern: string): Promise<string[]> {
+    if (pattern && !this.fieldCache.has(pattern)) {
+      try {
+        const fields = await this.http.get<string[] | undefined>(
```

Comment:
No valid cases, except for bugs. Would you prefer I narrow this to `string[]` to be clearer about the expected outcome?

Reply:
Yep, I'd rather keep only what we really expect. Otherwise reader has to go to the route definition and see when `undefined` can be returned to understand all the flows (as I did).